### PR TITLE
Update the RyuJIT IR documentation.

### PR DESCRIPTION
**Pull request by pgavlin**
The most important parts of these changes update the description of the
IR to cover HIR, LIR, and the differences between the two. Other updates
include some minor restructuring, typo fixes, rephrasings, and JIT dump
refreshes.

### CODE REVIEW

**Comment by pgavlin** [comment]
@briansull @BruceForstall @CarolEidt @dotnet/jit-contrib PTAL

**Comment by pgavlin** [comment]
Note that I hand-edited the JIT dump examples. These should really be refreshed from the original sources (especially in order to pick up the recent node name changes).

File: Documentation/botr/ryujit-overview.md [hunk 1]
```diff
@@ -243,11 +237,12 @@ As the JIT has evolved, changes have been made to improve the ability to reason
...
+* Elimination of “comma” nodes (`GT_COMMA`). These nodes are introduced for convenience during importation, during which a single tree is constructed at a time, and not incorporated into the statement list until it is completed. When it is necessary, for example, to store a partially-constructed tree into a temporary variable, a `GT_COMMA` node is used to link it into the tree. However, in later phases, these comma nodes are an impediment to analysis, and thus are eliminated.
   * In some cases, it is not possible to fully extract the tree into a separate statement, due to execution order dependencies. In these cases, an “embedded” statement is created. While these are conceptually very similar to the `GT_COMMA` nodes, they do not masquerade as expressions.
 * Elimination of “QMark” (`GT_QMARK`/`GT_COLON`) nodes is actually done at the end of morphing, long before the current rationalization phase. The presence of these nodes made analyses (especially dataflow) overly complex.
+* Elimiation of statements. Without statements, the execution order of a basic block's contents is fully defined by the `gtNext`/`gtPrev` links between `GenTree` nodes.
```
Comment (mikedn) [inline Documentation/botr/ryujit-overview.md]: Elimination

File: Documentation/botr/ryujit-overview.md [hunk 2]
```diff
@@ -28,45 +28,39 @@ RyuJIT provides the just in time compilation service for the .NET runtime. The r
...
+Both HIR and LIR blocks are composed of `GenTree` nodes that define the operations performed by the block. A `GenTree` node may consume some number of operands and may produce a singly-defined, at-most-singly-used value as a result (an SDSU or "tree" temp). Defs of SDSU temps are represented by `GenTree` nodes themselves, and uses are represented by edges from the using node to the defining node. Furthermore, SDSU temps defined in one block may not be used in a different block. In cases where a value must be multiply-defined, multiply-used, or defined in one block and used in another, the IR provides another class of temporary: the local var. Local vars are defined by assignment nodes in HIR or store nodes in LIR, and are used by local var nodes in both forms.
```
Comment (pgavlin) [inline Documentation/botr/ryujit-overview.md]: It certainly allows for multiple-register returns, but those registers logically comprise a single temp.

File: Documentation/botr/ryujit-overview.md [hunk 3]
```diff
@@ -103,22 +97,22 @@ The JIT is primarily concerned with “primitive” types, i.e. integers, refere
...
 
 The liveness analysis determines the set of defs, as well as the uses that are upward exposed, for each block. It then propagates the liveness information. The result of the analysis is captured in the following:
 
 * The live-in and live-out sets are captured in the `bbLiveIn` and `bbLiveOut` fields of the `BasicBlock`.
-* The `GTF_VAR_DEF` flag is set on a lvlVar `GenTree` node that is a definition.
+* The `GTF_VAR_DEF` flag is set on a lclVar `GenTree` node that is a definition.
 * The `GTF_VAR_USEASG` flag is set (in addition to the `GTF_VAR_DEF` flag) for the target of an update (e.g. +=).
-* The `GTF_VAR_USEDEF` is set on the target of an assignment of a binary operator with the same lvlVar as an operand.
+* The `GTF_VAR_USEDEF` is set on the target of an assignment of a binary operator with the same lclVar as an operand.
```
Comment (CarolEidt) [inline Documentation/botr/ryujit-overview.md]: You can delete this now :-)

Comment (CarolEidt) [inline Documentation/botr/ryujit-overview.md]: (And wow, how many times did I miss-spell `lclVar`?!)

File: Documentation/botr/ryujit-overview.md [hunk 4]
```diff
@@ -391,12 +413,12 @@ Post-conditions:
...
+* Determine the frame layout – allocating space on the frame for any lclVars that are not fully enregistered, as well as any spill temps required for spilling non-lclVar expressions.
 * For each `BasicBlock`, in layout order, and each `GenTree` node in the block, in execution order:
   * If the node is “contained” (i.e. its operation is subsumed by a parent node), do nothing.
   * Otherwise, “consume” all the register operands of the node.
-    * This updates the liveness information (i.e. marking a lvlVar as dead if this is the last use), and performs any needed copies.
-    * This must be done in correct execution order, obeying any reverse flags (GTF_REVERSE_OPS) on the operands, so that register conflicts are handled properly.
+    * This updates the liveness information (i.e. marking a lclVar as dead if this is the last use), and performs any needed copies.
+    * This must be done in "spill order" so that any spill/restore code inserted by the register allocator to resovle register conflicts is generated in the correct order. "
```
Comment (CarolEidt) [inline Documentation/botr/ryujit-overview.md]: resovle => resolve

File: Documentation/botr/ryujit-overview.md [hunk 5]
```diff
@@ -415,14 +437,14 @@ There are several properties of the IR that are valid only during (or after) spe
...
-    * Commutative operators, which can have the `GTF_REVERSE_OPS` flag set to indicate that op2 should be evaluated before op1.
-    * Assignments, which can also have the `GTF_REVERSE_OPS` flag set to indicate that the rhs (op2) should be evaluated before the target address (if any) on the lhs (op1) is evaluated. This can only be done if there are no side-effects in the expression for the lhs.
+    * Binary operators, which can have the `GTF_REVERSE_OPS` flag set to indicate that the RHS (`gtOp2`) should be evaluated before the LHS (`gtOp1`).
+    * Dynamically-sized block copy nodes, which can have `gtEvalSizeFirst` set to `true` to indicate that their `gtDynamicSize` tree should be evaluated before executing their other operands.
 * Rationalization
-  * All `GT_COMMA` nodes are split into separate statements, which may be embedded in other statements in execution order.
   * All `GT_ASG` trees are transformed into `GT_STORE` variants (e.g. `GT_STORE_LCL_VAR`).
   * All `GT_ADDR` nodes are eliminated (e.g. with `GT_LCL_VAR_ADDR`).
+  * All `GT_COMMA` and `GT_STMT` nodes are removed and their consituent nodes linked into execution order.
```
Comment (CarolEidt) [inline Documentation/botr/ryujit-overview.md]: consituent => constituent

**Pull request merged**